compiler/src/dmd/statementsem.d

```diff
@@ -806,6 +801,19 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
                             auto tf = fparam.type.nextOf().isTypeFunction();
                             foreachParamCount = tf.parameterList.length;
                             foundMismatch = true;
+
+                            // Mutability check
+                            if (fs.aggr && fs.aggr.type && fd.type && fs.aggr.type.isConst() && !fd.type.isConst())
```

Comment:
This bug can be triggered by *any* qualifier mismatch, and the qualifiers that trigger it may appear on either the `opApply` method, the `foreach` variable, or both. Checking specifically for `const` on the `foreach` variable and not on the `opApply` method is a band-aid that fixes the example in the bug report but leaves the underlying problem unsolved.